rpc/client/eventstream/eventstream.go

```diff
@@ -0,0 +1,188 @@
+	waitTime   time.Duration          // the long-polling interval
+	client     Client
+}
+
+// New constructs a new stream for the given query and options.
+// If opts == nil, the stream uses default values as described by
+// StreamOptions.  This function will panic if cli == nil.
+func New(cli Client, query string, opts *StreamOptions) *Stream {
+	if cli == nil {
+		panic("eventstream: nil client")
+	}
+	return &Stream{
+		filter:     &coretypes.EventFilter{Query: query},
+		batchSize:  opts.batchSize(),
+		newestSeen: opts.resumeFrom(),
+		waitTime:   opts.waitTime(),
+		client:     cli,
+	}
+}
+
+// Run polls the service for events matching the query, and calls accept for
+// each such event. Run handles pagination transparently, and delivers events
+// to accept in order of publication.
+//
+// Run continues until ctx ends or accept reports an error.  If accept returns
+// ErrStopRunning, Run returns nil; otherwise Run returns the error reported by
+// accept or ctx.  Run also returns an error if the server reports an error
+// from the Events method.
+//
+// If the stream falls behind the event log on the server, Run will stop and
+// report an error of concrete type *MissedItemsError.  Call Reset to reset the
+// stream to the head of the log, and call Run again to resume.
+func (s *Stream) Run(ctx context.Context, accept func(*coretypes.EventItem) error) error {
+	after := s.newestSeen // request (only) items after this
+	var pageCursor string // if non-empty, page through items before this
+	for {
+		var items []*coretypes.EventItem
+
+		// Fetch another batch of events from the service.
+		for {
+			rsp, err := s.client.Events(ctx, &coretypes.RequestEvents{
+				Filter:   s.filter,
+				MaxItems: s.batchSize,
+				After:    after,
+				Before:   pageCursor,
+				WaitTime: s.waitTime,
+			})
+			if err != nil {
+				return err
+			}
+
+			// If the oldest item in the log is newer than our most recent item,
+			// it means we might have missed some events matching our query.
+			if s.newestSeen != "" && s.newestSeen < rsp.Oldest {
+				return &MissedItemsError{
+					Query:         s.filter.Query,
+					NewestSeen:    s.newestSeen,
+					OldestPresent: rsp.Oldest,
+				}
+			}
+			items = append(items, rsp.Items...)
+
+			if rsp.More {
+				// There are more results matching this request, leave the baseline
+				// where it is and set the page cursor so that subsequent requests
+				// will get the next chunk.
+				pageCursor = items[len(items)-1].Cursor
+			} else if len(items) != 0 {
+				// We got everything so far; skip the base past what we've seen.
+				pageCursor = ""
+				after = items[0].Cursor
+				break
+			}
+		}
+
+		// Deliver events from the current batch to the receiver.  We visit the
+		// batch in reverse order so the receiver sees them in forward order.
+		for i := len(items) - 1; i >= 0; i-- {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			default:
+			}
```

Comment:
I always do `if err := ctx.Err(); err != nil {return err}` in these cases

Reply:
Ah, yeah. Once upon a time `ctx.Err()` was not guaranteed to have a meaningful value until after termination. I still have that reflex. Will update.